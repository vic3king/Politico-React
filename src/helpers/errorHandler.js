const errorHandler = obj => {
  let error;

  if (typeof obj === 'string') {
    error = obj;
  }

  if (typeof obj === 'object') {
    Object.values(obj).forEach(err => {
      error = err;
    });
  }

  return error;
};

export default errorHandler;
