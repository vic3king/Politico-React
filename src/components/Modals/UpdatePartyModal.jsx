import React, { Component } from 'react';
import { notify } from 'react-notify-toast';
import Modal from '../shared/Modal/Modal';
import Input from '../shared/InputFields/Input';
import Button from '../shared/Buttons/Button';
import Header from '../shared/Header/Header';
import Parties from '../../services/parties';
import Loader from '../shared/Loader/Loader';
import errorHandler from '../../helpers/errorHandler';

class UpdatePartyModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formDetails: {},
      loading: false,
    };
  }

  onInputChange = event => {
    const { formDetails } = this.state;
    formDetails[event.target.id] = event.target.value;
    this.setState({ formDetails });
  };

  handleUpdate = async () => {
    this.setState({ loading: true });
    const { partyId, updatePartiesName, hide } = this.props;
    const { formDetails } = this.state;

    const party = await Parties.updateParty(partyId, formDetails);

    if (party.status >= 400) {
      this.setState({ loading: false });
      notify.show(errorHandler(party.error), 'error');
    }

    if (party.status === 200) {
      this.setState({ loading: false });
      updatePartiesName(partyId, party.data.name);
      notify.show(party.message);
      hide();
    }
  };

  hide = e => {
    e.preventDefault();
    const { hide } = this.props;
    hide();
  };

  render() {
    const { formDetails, loading } = this.state;
    return (
      <React.Fragment>
        {loading && <Loader />}
        <Modal
          Header={<Header text="Update Party" />}
          Select={
            <Input
              type="text"
              id="name"
              placeholder="update party name"
              value={formDetails.name}
              onChange={this.onInputChange}
              required
            />
          }
          ButtonOne={
            <Button
              id="party-btn"
              type="submit"
              className="cte-off"
              value="Update"
              onClick={this.handleUpdate}
            />
          }
          CloseModal={
            <button type="button" id="btn" onClick={this.hide}>
              X
            </button>
          }
        />
      </React.Fragment>
    );
  }
}

export default UpdatePartyModal;
