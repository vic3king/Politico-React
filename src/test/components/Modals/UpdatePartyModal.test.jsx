import React from 'react';
import { shallow } from 'enzyme';
import { notify } from 'react-notify-toast';
import UpdatePartyModal from '../../../components/Modals/UpdatePartyModal';
import Party from '../../../services/parties';
import errorHandler from '../../../helpers/errorHandler';

jest.mock('react-notify-toast');
jest.mock('../../../services/parties');

const hide = jest.fn();
const updatePartiesName = jest.fn();
const partyId = 1;

let wrapper;
describe('InterestFormModal component', () => {
  global.fetch = jest.fn();
  global.localStorage.setItem('user', '{}');

  beforeEach(() => {
    wrapper = shallow(
      <UpdatePartyModal
        updatePartiesName={updatePartiesName}
        hide={hide}
        partyId={partyId}
      />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('hide method', () => {
    wrapper = shallow(
      <UpdatePartyModal
        updatePartiesName={updatePartiesName}
        hide={hide}
        partyId={partyId}
      />
    );
    const instance = wrapper.instance();
    let event;
    it('should call preventDefault on event', async () => {
      event = {
        preventDefault: jest.fn(),
      };

      await instance.hide(event);
      expect(event.preventDefault).toBeCalled();
    });
  });

  describe('onInputChange method', () => {
    it('should update formData state ', () => {
      const instance = wrapper.instance();
      const event = {
        preventDefault: jest.fn(),
        target: { id: 'office' },
      };
      const formDetails = {
        name: 'test',
      };
      instance.setState({
        formDetails,
      });

      instance.onInputChange(event);

      expect(instance.state.formDetails).toEqual({
        name: 'test',
      });
    });
  });

  describe('handleUpdate method', () => {
    let instance;

    describe('Api call success', () => {
      beforeAll(() => {
        Party.updateParty = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 200,
            message: 'toast message',
            data: {
              name: 'test',
            },
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
      });

      it('should call toast the right message on api call success', async () => {
        await instance.handleUpdate();
        expect(notify.show).toBeCalledWith('toast message');
      });
      it('should set loading state to false after successfull api call is made', async () => {
        await instance.handleUpdate();
        expect(instance.state.loading).toBeFalsy();
      });
      it('should call the updatePartiesName method with id from api call', async () => {
        await instance.handleUpdate();
        expect(updatePartiesName).toBeCalledWith(1, 'test');
      });
    });

    describe('Api call 400 errors', () => {
      beforeAll(() => {
        Party.updateParty = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 400,
            error: 'validation error',
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
      });

      it('should set loading state to false on validation error', async () => {
        await instance.handleUpdate();
        expect(instance.state.loading).toBeFalsy();
      });

      it('should toast the error message', async () => {
        await instance.handleUpdate();
        expect(notify.show).toBeCalledWith(
          errorHandler('validation error'),
          'error'
        );
      });
    });
  });
});
