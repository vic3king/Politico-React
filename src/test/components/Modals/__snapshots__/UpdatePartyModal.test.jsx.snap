// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterestFormModal component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UpdatePartyModal
    hide={[MockFunction]}
    partyId={1}
    updatePartiesName={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        false,
        <Modal
          ButtonOne={
            <Button
              className="cte-off"
              id="party-btn"
              onClick={[Function]}
              type="submit"
              value="Update"
            />
          }
          CloseModal={
            <button
              id="btn"
              onClick={[Function]}
              type="button"
            >
              X
            </button>
          }
          Header={
            <Header
              text="Update Party"
            />
          }
          InputOne=""
          InputThree=""
          InputTwo=""
          Select={
            <Input
              id="name"
              onChange={[Function]}
              placeholder="update party name"
              required={true}
              type="text"
              value=""
            />
          }
          SpanFour=""
          SpanOne=""
          SpanThree=""
          SpanTwo=""
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "ButtonOne": <Button
            className="cte-off"
            id="party-btn"
            onClick={[Function]}
            type="submit"
            value="Update"
          />,
          "CloseModal": <button
            id="btn"
            onClick={[Function]}
            type="button"
          >
            X
          </button>,
          "Header": <Header
            text="Update Party"
          />,
          "InputOne": "",
          "InputThree": "",
          "InputTwo": "",
          "Select": <Input
            id="name"
            onChange={[Function]}
            placeholder="update party name"
            required={true}
            type="text"
            value=""
          />,
          "SpanFour": "",
          "SpanOne": "",
          "SpanThree": "",
          "SpanTwo": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          false,
          <Modal
            ButtonOne={
              <Button
                className="cte-off"
                id="party-btn"
                onClick={[Function]}
                type="submit"
                value="Update"
              />
            }
            CloseModal={
              <button
                id="btn"
                onClick={[Function]}
                type="button"
              >
                X
              </button>
            }
            Header={
              <Header
                text="Update Party"
              />
            }
            InputOne=""
            InputThree=""
            InputTwo=""
            Select={
              <Input
                id="name"
                onChange={[Function]}
                placeholder="update party name"
                required={true}
                type="text"
                value=""
              />
            }
            SpanFour=""
            SpanOne=""
            SpanThree=""
            SpanTwo=""
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "ButtonOne": <Button
              className="cte-off"
              id="party-btn"
              onClick={[Function]}
              type="submit"
              value="Update"
            />,
            "CloseModal": <button
              id="btn"
              onClick={[Function]}
              type="button"
            >
              X
            </button>,
            "Header": <Header
              text="Update Party"
            />,
            "InputOne": "",
            "InputThree": "",
            "InputTwo": "",
            "Select": <Input
              id="name"
              onChange={[Function]}
              placeholder="update party name"
              required={true}
              type="text"
              value=""
            />,
            "SpanFour": "",
            "SpanOne": "",
            "SpanThree": "",
            "SpanTwo": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
